proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,305 @@
+- *many to one* - kroxy presents many downstream clusters which route to a single upstream - this would support a multitenant use-cases 
+
+(There's a possiblity of *one-to-many* which a single downstream routes to several upstreams as if they were one, which might be useful for presenting several
+kafka clusters as if they were one, however, supporting transactions across two clusters would be difficult.  We consider won't this use-case futrther).
+
+## Current situation
+
+Currently Kroxyilicious binds a single port to the downstream and connects to a single upstream broker.  There is no support for clusters comprised of more
+then one broker.
+
+## Motivation
+
+1. Ability for Kroxylicious to expose multiple physical kafka clusters.
+1. Ability for Kroxylicious to support clusters formed of more than one broker.
+1. Ability for Kroxylicious to provide a building block to support multi-tenancy.
+
+## Proposal
+
+The proposal will change some existing concepts and introduce some new ones.  These are discussed first.  Then, we talk about the changes in high level responsibilities.
+
+### Key Objects
+
+
+![alternative text](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/UML/uml.puml)
+
+
+
+
+
+#### Endpoints
+
+Currently `KafkaProxy` uses the `ProxyConfig` object to get a single address to bind to.  
+
+* For the non-TLS case, multiple endpoints are required to support >1 brokers. 
+* For TLS case, SNI gives a mechanism to route the traffic to a >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.
+
+`ProxyConfig` will be replaced by an list of _Endpoint_ objects.  
+
+```
+- name: endpoint name
+  address: bind address/port
+  type: PLAIN|TLS
+```
+
+#### Downstream Clusters
+
+Conceptually the _downstream cluster_ represents the cluster that the kafka client connects to.  The downstream clusters list specifies
+all the kafka clustsers are being presented by Kroxylicious. 
+
+A _downstream clustser_ references exactly one _upstream cluster_. 
+
+A _downstream cluster_ may reference a `filter chain`.  This provides zero or more filters that the RPCs will pass through before arriving at the upstream cluster.
+
+A _downstream cluster_ enumerates the downstream brokers that comprise the downstream cluster.
+
+##### Downstream Brokers
+
+Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
+for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.
+
+Each downstream broker references exactly one upstream broker which must exist within the referenced upstream cluster.  It is an error if there is not a
+one-to-one correspondence between downstream and upstream brokers.
```

Comment:
Does this hold true in the multi tenant use case? @robobario and I are imagining providing N `virtual` clusters compose of `3` brokers each being serviced by a **single** `physical` cluster composed of `6` brokers. 

We could establish fixed mappings from `virtual` brokers to `physical` brokers initially but when we use cruse control to rebalance topics across the `6 physical` brokers those fixed mapping no longer hold.  As the partitions could be moved to the brokers not included in the one to one mapping. We potentially need to be able to map multiple `physical` brokers to a single `virtual` broker. 


We may not need to support this the MVP but any design we come up with should be open to extending in this direction. In the MVP we could enforce that all `virtual` clusters have the same broker counts as the `physical` clusters.

Reply:
> Does this hold true in the multi tenant use case? @robobario and I are imagining providing N `virtual` clusters compose of `3` brokers each being serviced by a **single** `physical` cluster composed of `6` brokers.

Can that work?   I had been working under the (unspoken) assumption that a virtual cluster had n virtual brokers where n matched the number of the physical brokers on the cluster. It is possible I'm the outlier here.

> We could establish fixed mappings from virtual brokers to physical brokers initially but when we use cruise control to rebalance topics across the 6 physical brokers those fixed mapping no longer hold.

Exactly.

> We potentially need to be able to map multiple physical brokers to a single virtual broker.

I'm not sure that could work.  When a topic is created, the kafka client decides which broker each partition/replicas go.   If there's not a one to one mapping between virtual broker and physical ones I think this would break down.

It might be possible to have kroxy expose only those brokers hosting partitions where the tenant has partitions but this does seen to bring a fair amount of extra complexity.

What's the advantage for exposing only a subset of the physical brokers?

Reply:
On reflection I don't think `N physical` brokers to `1 virtual` broker doesn't work and is unnecessary but I think we do need to support a subset of brokers (at some point).

The use-case I am thinking of is something like offering small clusters on a much larger shared cluster, in the tens or hundreds of brokers. I don't see it making any realistic sense to expose hundreds of brokers through the proxy to a single `virtual` cluster. If we expand to that point having a static map of `virtual` clusters to `physical` brokers becomes very difficult to manage. 

For the record I'm not saying we need to fix this during MVP, just that we need to keep this end goal in mind and keep options open that make that support easier.

Reply:
> For the record I'm not saying we need to fix this during MVP, just that we need to keep this end goal in mind and keep options open that make that support easier.
+1

I'll call that out as a goal.

Reply:
Done.